### PR TITLE
Add payment contract

### PR DESCRIPTION
**Pull request by volodymyr-basiuk**

### CODE REVIEW

File: contracts/examples/PayExample.sol [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    event Payment(uint256 indexed issuerId, string paymentId, uint256 schemaHash);
+
+    constructor() Ownable(_msgSender()) { }
+
+    function setPaymentValue(uint256 issuerId, uint256 schemaHash, uint256 value) public onlyOwner {
+        ValueToPay[keccak256(abi.encode(issuerId, schemaHash))] = value;
+    }
+
+    function pay(string calldata paymentId, uint256 issuerId, uint256 schemaHash) public payable {
+        uint256 requiredValue = ValueToPay[keccak256(abi.encode(issuerId, schemaHash))];
```
Comment (daveroga) [inline contracts/examples/PayExample.sol:20]: As discussed in a call, I think we should require here only 1 payment for the same (paymentId, issuerId)

**Review by daveroga — CHANGES_REQUESTED** [review]

File: contracts/examples/PayExample.sol (hunk 1 above)
Comment (volodymyr-basiuk) [reply contracts/examples/PayExample.sol:25]: added

**Review by daveroga — APPROVED** [review]

File: contracts/examples/PayExample.sol [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        }
+        payments[payment] = true;
+        emit Payment(issuerId, paymentId, schemaHash);
+    }
+
+    function isPaymentDone(string calldata paymentId, uint256 issuerId) public view returns (bool) {
+        return payments[keccak256(abi.encode(issuerId, paymentId))];
+    }
+
+    function withdraw(uint amount) public onlyOwner {
```
Comment (AndriianChestnykh) [inline contracts/examples/PayExample.sol:46]: May be better not to use `amount` at all.

File: contracts/examples/PayExample.sol [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        require(amount <= address(this).balance);
+        (bool sent,) = owner().call{ value:  amount }("");
+        require(sent, "Failed to withdraw");
+    }
+
+    function getContractBalance() public view returns(uint){
```
Comment (AndriianChestnykh) [inline contracts/examples/PayExample.sol:52]: This is not mandatory. Provided by blockchain infrastructure.

File: contracts/examples/PayExample.sol [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+     /**
+     * @dev mapping of paymentRequest - keccak256(abi.encode(issuerId, paymentId)) => bool
+     */
+    mapping (bytes32 paymentRequest => bool isPaid) public payments;
+
+    event Payment(uint256 indexed issuerId, string indexed paymentId, uint256 schemaHash);
+
+    error PaymentError(string message);
+
```
Comment (daveroga) [inline contracts/examples/PayExample.sol:19]:
```suggestion
error WithdrawError(string message);
```

File: contracts/examples/PayExample.sol [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    function isPaymentDone(string calldata paymentId, uint256 issuerId) public view returns (bool) {
+        return payments[keccak256(abi.encode(issuerId, paymentId))];
+    }
+
+    function withdraw() public onlyOwner {
+        (bool sent,) = owner().call{ value:  address(this).balance }("");
+        require(sent, "Failed to withdraw");
+    }
```
Comment (daveroga) [inline contracts/examples/PayExample.sol:49]: Control balance to withdraw and `WithdrawError` like `PaymentError`

```suggestion
    function withdraw() public onlyOwner {
        if (address(this).balance == 0) {
            revert WithdrawError("There is no balance to witdraw");
        }
        (bool sent,) = owner().call{ value:  address(this).balance }("");
        if (!sent) {
            revert WithdrawError("Failed to withdraw");
        }
    }
```

File: contracts/examples/PayExample.sol (hunk 2 above)
Comment (volodymyr-basiuk) [reply contracts/examples/PayExample.sol]: removed `amount`

File: contracts/examples/PayExample.sol [hunk 6]
```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.20;
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+
+contract PayExample is Ownable {
+    /**
+     * @dev mapping of keccak256(abi.encode(issuerId, schemaHash)) => value
+     */
+    mapping (bytes32 => uint256) private valueToPay;
```
[the remark this answers is not in the document]
Comment (volodymyr-basiuk) [reply contracts/examples/PayExample.sol]: added naming to maps

File: contracts/examples/PayExample.sol [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+contract PayExample is Ownable {
+    /**
+     * @dev mapping of paymentValueId - keccak256(abi.encode(issuerId, schemaHash)) => value
+     */
+    mapping (bytes32 paymentValueId => uint256 value) private valueToPay;
+
+     /**
+     * @dev mapping of paymentRequest - keccak256(abi.encode(issuerId, paymentId)) => bool
+     */
+    mapping (bytes32 paymentRequest => bool isPaid) public payments;
```
Comment (AndriianChestnykh) [inline contracts/examples/PayExample.sol:14]: May be `paymentRequestId` is better than `paymentRequest`

**Review by AndriianChestnykh — APPROVED** [review]
Only one comment. The rest looks ok

File: contracts/examples/PayExample.sol (hunk 7 above)
Comment (volodymyr-basiuk) [reply contracts/examples/PayExample.sol]: renamed

**Review by AndriianChestnykh — APPROVED** [review]

File: hardhat.config.ts [hunk 8]
```diff
@@ -37,6 +37,16 @@ const config: HardhatUserConfig = {
     //   url: `${process.env.AMOY_RPC_URL}`,
     //   accounts: [`0x${process.env.AMOY_PRIVATE_KEY}`]
     // },
+    // 'linea-sepolia': {
+    //   chainId: 59141,
+    //   url: `${process.env.LINEA_SEPOLIA_RPC_URL}`,
+    //   accounts: [`0x${process.env.LINEA_SEPOLIA_PRIVATE_KEY}`]
+    // },
+    linea: {
```
Comment (vmidyllic) [inline hardhat.config.ts]: comment it?

**Review by vmidyllic — APPROVED** [review]

**Pull request merged**